### ISSUE TITLE
Allow importing multi-file GGUF models

### ISSUE THREAD

**Issue by jmorganca** [ollama/ollama#5245]
### What is the issue?

Currently Ollama can [import GGUF files](https://github.com/ollama/ollama/blob/main/docs/import.md). However, larger models are sometimes split into separate files. Ollama should support loading multiple GGUF files similar to loading safetensor files.

**Comment by gsoul** [comment]
Just in case someone would find this issue, like I did a few weeks ago, without knowing any workaround. Currently probably one of the easiest ways import multifile gguf into Ollama would be to:
1. Download pre-compiled binaries of llama.cpp: https://github.com/ggerganov/llama.cpp/releases (or install according to their manual)
2. run command:
`./llama-gguf-split --merge mymodel-00001-of-00002.gguf out_file_name.gguf`
for example
`./llama-gguf-split --merge Mistral-Large-Instruct-2407-IQ4_XS-00001-of-00002.gguf outfile.gguf`

Hope this will help somebody.

**Comment by nauen** [comment]
> Just in case someone would find this issue, like I did a few weeks ago, without knowing any workaround. Currently probably one of the easiest ways import multifile gguf into Ollama would be to:
> 
> 1. Download pre-compiled binaries of llama.cpp: https://github.com/ggerganov/llama.cpp/releases (or install according to their manual)
> 2. run command:
>    `./llama-gguf-split --merge mymodel-00001-of-00002.gguf out_file_name.gguf`
>    for example
>    `./llama-gguf-split --merge Mistral-Large-Instruct-2407-IQ4_XS-00001-of-00002.gguf outfile.gguf`
> 
> Hope this will help somebody.

yes it does <3

**Comment by werruww** [comment]
Does Obama support fragmented models? Important: They must be merged before running in Olama by modelfile

**Comment by werruww** [comment]
lama cpp and llama cpp python can  run multy part


ollama can run it???????????????

**Comment by werruww** [comment]
/content# ollama run hf.co/goodasdgood/dracarys2-72b-instruct
pulling manifest 
Error: pull model manifest: 400: The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: https://github.com/ollama/ollama/issues/5245
/content#

**Comment by mitchross** [comment]
https://x.com/reach_vb/status/1846545312548360319

**Comment by ahmetkca** [comment]
Having this issue with `Qwen/Qwen2.5-Coder-32B-Instruct-GGUF:Q8_0`

**Comment by rar0n** [comment]
> Having this issue with `Qwen/Qwen2.5-Coder-32B-Instruct-GGUF:Q8_0`

Try what gsoul above suggests!
I just did and it worked for qwen2.5-coder-14b-instruct-q5_k_m-00001-of-00002.gguf. Didn't have to specify any other input files.
(Thanks gsoul!)

Still, it'd be nice if ollama could do this natively ofc.

**Comment by Kamayuq** [comment]
Even though the workaround works it is a very cumbersome especially if you have to update the model. And you also have to create a manifest AFAIK. Ollama should really do this locally completely transparent for the user.

**Comment by DrewGalbraith** [comment]
> Just in case someone would find this issue, like I did a few weeks ago, without knowing any workaround. Currently probably one of the easiest ways import multifile gguf into Ollama would be to:
> 
> 1. Download pre-compiled binaries of llama.cpp: https://github.com/ggerganov/llama.cpp/releases (or install according to their manual)
> 2. run command:
>    `./llama-gguf-split --merge mymodel-00001-of-00002.gguf out_file_name.gguf`
>    for example
>    `./llama-gguf-split --merge Mistral-Large-Instruct-2407-IQ4_XS-00001-of-00002.gguf outfile.gguf`
> 
> Hope this will help somebody.

For anyone else wondering how this merges all the other parts of the model or if you have to run this command for each split of the file to merge it into the new one or something, this [reddit post](https://www.reddit.com/r/LocalLLaMA/comments/1cf6n18/comment/l1o0opp/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) laid it out. The gguf-split utility simply figures out the rest of the shards to merge together given the name of the first. I assume it requires they be names basically the same with only numbers deviating.

**Comment by AlgorithmicKing** [comment]
> In case you want a way to easily handle this, add this to your .bashrc or .zshrc file:
> 
> ```
> add_gguf() {
>     set -x
>     args="$@"
>     cd /Volumes/MP44/Ollama
>     s="${args/ollama run /}"
>     IFS=: read -r ggufUrl quant <<< "$s"
>     IFS=/ read -r ggufD ggufU gguf <<< "$ggufUrl"
>     echo "[$ggufUrl] [$gguf] [$quant]"
>     ggufN="${gguf/-GGUF/}"
>     wget "${ggufUrl}/resolve/main/${ggufN}-${quant}/${ggufN}-${quant}-00001-of-00002.gguf"
>     wget "${ggufUrl}/resolve/main/${ggufN}-${quant}/${ggufN}-${quant}-00002-of-00002.gguf"
>     bash -c "llama-gguf-split --merge ${ggufN}-${quant}-00001-of-00002.gguf ${ggufN}-${quant}.guff"
>     echo "FROM /Volumes/MP44/Ollama/${ggufN}-${quant}.guff" > "${ggufN}-${quant}.model"
>     bash -c "ollama create ${ggufN}-${quant} -f ${ggufN}-${quant}.model"
>     bash -c "rm -f ${ggufN}*"
> }
> ```
> 
> Usage example: add_gguf hf.co/bartowski/EVA-LLaMA-3.33-70B-v0.1-GGUF:Q6_K
> 
> It also handles the edge case where you've copied the full command out of hugging face (strips "ollama run" out): add_gguf ollama run hf.co/bartowski/EVA-LLaMA-3.33-70B-v0.1-GGUF:Q6_K
> 
> What it does:
> 
> 1. Changes directory to /Volumes/MP44/Ollama (you can customize this for your own external drive, or wherever you want to temporarily store the downloaded .guff files)
> 2. Downloads the two .gguf files for the model using wget
> 3. Executes llama-gguf-split to merge the .gguf files into a single .guff
> 4. Creates the .model file
> 5. Executes ollama create using the .model file
> 6. Cleans up the downloaded files
> 
> After executing the command, you can run your model like this: ollama run EVA-LLaMA-3.33-70B-v0.1-Q6_K
> 
> Verify the model information by entering: /show info
> 
> ```
>   Model
>     architecture        llama
>     parameters          70.6B
>     context length      131072
>     embedding length    8192
>     quantization        Q6_K
> ```
> 
> To stop, enter: /bye

where are "bashrc or .zshrc file" located?

**Comment by BornSupercharged** [comment]
@AlgorithmicKing in your user directory, i.e. ~/.bashrc

**Comment by ngxson** [comment]
Upstream llama.cpp added a new API called `llama_model_load_from_splits` that may help implementing this function on ollama. Let's hope that they will work on this on the next version!

**Comment by mattapperson** [comment]
Here is an updated version of @BornSupercharged's script that:

- Accounts for splits that have an unknown number of parts. 
- Works on Mac
- Does some pre-flight checks to ensure everything works as expected
- Downloads files in parallel for faster pulls
- Returns the user to their previous directory when done

```
add_gguf() {
    set -x
    args="$@"
 
    # Check if ollama is installed and get version
    version_output=$(ollama --version 2>&1)
    
    # Check if the output contains "Warning"
    if echo "$version_output" | grep -q "Warning"; then
        echo "Error: Ollama version check failed with warning: $version_output"
        echo "Please be sure that ollama client and server are installed, running, and on the same version."
        return 1
    fi

    # Check if llama-gguf-split is available in PATH
    if ! command -v llama-gguf-split &> /dev/null; then
        echo "Error: llama-gguf-split not found in PATH"
        echo "Please install the latest version of llama.cpp first"
        return 1
    fi

    # Save current directory to restore it later
    current_dir=$(pwd)
    
    # Use a more standard Mac location
    OLLAMA_DIR="$HOME/Library/Ollama"
    mkdir -p "$OLLAMA_DIR"
    cd "$OLLAMA_DIR" || return
    s="${args/ollama run /}"
    IFS=: read -r ggufUrl quant <<< "$s"
    IFS=/ read -r ggufD ggufU gguf <<< "$ggufUrl"
    echo "[$ggufUrl] [$gguf] [$quant]"
    ggufN="${gguf/-GGUF/}"

    git clone --no-checkout --depth 1 "https://${ggufUrl}"
    cd "${gguf}/.git"
    files=$(git ls-tree --full-name --name-only -r HEAD)
    cd ".."

    
    # Create array to store wget commands
    wget_commands=()
    
    while IFS= read -r file; do
      if [[ "$file" == *"$quant"* ]]; then
        if [[ -z "${firstFileName}" ]] && [[ "${file}" == *"01-of"* ]]; then
            firstFileName="${file}"
        fi
        if [[ -f "${file}" ]]; then
            echo "File ${file} already exists, skipping..."
            continue
        fi
        echo "Downloading https://${ggufUrl}/resolve/main/${file}"
        wget_commands+=("wget https://${ggufUrl}/resolve/main/${file}")
      fi
    done <<< "$files"
    
    # Execute all wget commands in parallel
    printf "%s\n" "${wget_commands[@]}" | xargs -P 8 -I {} bash -c "{}"


    bash -c "llama-gguf-split --merge ${firstFileName} ${ggufN}-${quant}.guff"
    echo "FROM ./${ggufN}-${quant}.guff" > "${ggufN}-${quant}.model"
    bash -c "ollama create ${ggufN}-${quant} -f ${ggufN}-${quant}.model"
    bash -c "rm -f ${ggufN}*"

    cd "$current_dir" || return
}
```

**Comment by sakthi-geek** [comment]
Why is this not natively supported yet? Is it being worked on for future updates?

**Comment by AlgorithmicKing** [comment]
> Please work on this guys <3 we love ollama and want to continue using it!
> 
> `ollama run hf.co/unsloth/DeepSeek-R1-GGUF:Q4_K_M`
> 
> pulling manifest **Error**: _pull model manifest: 400: The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: [#5245](https://github.com/ollama/ollama/issues/5245)_ (this page)

I 100% agree with you but if you want to pull the full version of deepseek r1 it is already on [ollama](https://ollama.com/library/[deepseek-r1:671b](https://ollama.com/library/deepseek-r1:671b)) you can run `ollama run deepseek-r1:671b`

**Comment by LeiHao0** [comment]
> > Please work on this guys <3 we love ollama and want to continue using it!
> > `ollama run hf.co/unsloth/DeepSeek-R1-GGUF:Q4_K_M`
> > pulling manifest **Error**: _pull model manifest: 400: The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: [#5245](https://github.com/ollama/ollama/issues/5245)_ (this page)
> 
> I 100% agree with you but if you want to pull the full version of deepseek r1 it is already on [ollama](https://ollama.com/library/%5Bdeepseek-r1:671b%5D(https://ollama.com/library/deepseek-r1:671b)) you can run `ollama run deepseek-r1:671b`

Will ollama considering unsloth/DeepSeek-R1-GGUF, the 1.58-bit + 2-bit Dynamic Quants?

**Comment by corticalstack** [comment]
Watching this space for solutions to run lower quantized versions of DeepSeek-R1 that mere mortals can self host, e.g. 1.58-bit + 2-bit Dynamic Quants?

**Comment by Tanote650** [comment]
Implementing this in Ollama and giving a large number of less experienced users access to the models would be great!

**Comment by LeiHao0** [comment]
## Good News:

I've successfully quantized and run the DeepSeek-R1 model at 1.58 bits on my M2 Ultra, using the instructions from https://unsloth.ai/blog/deepseekr1-dynamic. 

## Bad News: 

Higher quantization levels of 1.73 bits, 2.22 bits and 2.51 bits failed to run. More importantly, even the successful 1.58 bits 
 model generate nonsensical output despite multiple attempts.

<img width="577" alt="Image" src="https://github.com/user-attachments/assets/d68bf7fa-22f5-4845-9ed9-1d39cf51b98a" />

**Comment by corticalstack** [comment]
Hoping some smaller DeepSeek quants / multi-file GGUF support provided soon, given the HUGE interest in this model right now.

**Comment by LeiHao0** [comment]
> [@LeiHao0](https://github.com/LeiHao0) can you confirm you were able to run it with ollama? If so, can you share your Modelfile?

You need to merge these split model files into a single file, then you can load it with ollama.
details in here https://unsloth.ai/blog/deepseekr1-dynamic

`./llama.cpp/llama-gguf-split --merge \
  DeepSeek-R1-GGUF/DeepSeek-R1-UD-IQ1_S/DeepSeek-R1-UD-IQ1_S-00001-of-00003.gguf \
	merged_file.gguf`

**Comment by verygreen** [comment]
@LeiHao0 don't forget to also implement this context fix if you want for the output to actually run to its completion: https://github.com/ollama/ollama/issues/5975#issuecomment-2295330804 (obviously don't use 24k context unless you have gobs of video RAM).

But overall I am having bad results with this particular quant in ollama, even thinking tags don't appear and the model seems to be rambling on endlessly until eventually cutting out

**Comment by dmatora** [comment]
What about template section for DeepSeek Modelfile?
Doesn't it need one?

**Comment by dmatora** [comment]
Ok, found a tip at issue #8571 
```
FROM merged_file.gguf
TEMPLATE """{{- if .System }}{{ .System }}{{ end }}
{{- range $i, $_ := .Messages }}
{{- $last := eq (len (slice $.Messages $i)) 1}}
{{- if eq .Role "user" }}<｜User｜>{{ .Content }}
{{- else if eq .Role "assistant" }}<｜Assistant｜>{{ .Content }}{{- if not $last }}<｜end▁of▁sentence｜>{{- end }}
{{- end }}
{{- if and $last (ne .Role "assistant") }}<｜Assistant｜>{{- end }}
{{- end }}"""
PARAMETER stop <｜begin▁of▁sentence｜>
PARAMETER stop <｜end▁of▁sentence｜>
PARAMETER stop <｜User｜>
PARAMETER stop <｜Assistant｜>
```

**Comment by mistrjirka** [comment]
> > [@LeiHao0](https://github.com/LeiHao0) can you confirm you were able to run it with ollama? If so, can you share your Modelfile?
> 
> You need to merge these split model files into a single file, then you can load it with ollama. details in here https://unsloth.ai/blog/deepseekr1-dynamic
> 
> `./llama.cpp/llama-gguf-split --merge \ DeepSeek-R1-GGUF/DeepSeek-R1-UD-IQ1_S/DeepSeek-R1-UD-IQ1_S-00001-of-00003.gguf \ merged_file.gguf`

It seems that the ollama does not support IQ1 bit quantization. So for the 1bit quantization to work maybe some update to ollama is needed. But it is weird because llama.cpp supports the one bit quantization. The ollama errors out on wrong magic number.

**Comment by peng3502** [comment]
There are 10 files for DeepSeek
DeepSeek-R1-Q5_K_M-00001-of-00010.gguf  DeepSeek-R1-Q5_K_M-00005-of-00010.gguf  DeepSeek-R1-Q5_K_M-00009-of-00010.gguf
DeepSeek-R1-Q5_K_M-00002-of-00010.gguf  DeepSeek-R1-Q5_K_M-00006-of-00010.gguf  DeepSeek-R1-Q5_K_M-00010-of-00010.gguf
DeepSeek-R1-Q5_K_M-00003-of-00010.gguf  DeepSeek-R1-Q5_K_M-00007-of-00010.gguf
DeepSeek-R1-Q5_K_M-00004-of-00010.gguf  DeepSeek-R1-Q5_K_M-00008-of-00010.gguf

An error throwed out after llama merge.

llama-gguf-split --merge DeepSeek-R1-Q5_K_M-00001-of-00010.gguf DeepSeek-R1-Q5.gguf
gguf_merge: DeepSeek-R1-Q5_K_M-00001-of-00010.gguf -> DeepSeek-R1-Q5.gguf
terminate called after throwing an instance of 'std::__ios_failure'
  what():  basic_ios::clear: iostream error

Does any one else have a solution?

**Comment by MaoJianwei** [comment]
> Just in case someone would find this issue, like I did a few weeks ago, without knowing any workaround. Currently probably one of the easiest ways import multifile gguf into Ollama would be to:
> 
> 1. Download pre-compiled binaries of llama.cpp: https://github.com/ggerganov/llama.cpp/releases (or install according to their manual)
> 2. run command:
>    `./llama-gguf-split --merge mymodel-00001-of-00002.gguf out_file_name.gguf`
>    for example
>    `./llama-gguf-split --merge Mistral-Large-Instruct-2407-IQ4_XS-00001-of-00002.gguf outfile.gguf`
> 
> Hope this will help somebody.



This works for me! Thanks!



```
./llama-gguf-split --merge DeepSeek-R1-Distill-Qwen-32B-F16-00001-of-00002.gguf DeepSeek-R1-Distill-Qwen-32B-F16.gguf

gguf_merge: DeepSeek-R1-Distill-Qwen-32B-F16-00001-of-00002.gguf -> DeepSeek-R1-Distill-Qwen-32B-F16.gguf
gguf_merge: reading metadata DeepSeek-R1-Distill-Qwen-32B-F16-00001-of-00002.gguf done
gguf_merge: reading metadata DeepSeek-R1-Distill-Qwen-32B-F16-00002-of-00002.gguf done
gguf_merge: writing tensors DeepSeek-R1-Distill-Qwen-32B-F16-00001-of-00002.gguf done
gguf_merge: writing tensors DeepSeek-R1-Distill-Qwen-32B-F16-00002-of-00002.gguf done
gguf_merge: DeepSeek-R1-Distill-Qwen-32B-F16.gguf merged from 2 split with 771 tensors.

```

**Comment by thalesluoyx** [comment]
> ## Good News:
> I've successfully quantized and run the DeepSeek-R1 model at 1.58 bits on my M2 Ultra, using the instructions from https://unsloth.ai/blog/deepseekr1-dynamic.
> 
> ## Bad News:
> Higher quantization levels of 1.73 bits, 2.22 bits and 2.51 bits failed to run. More importantly, even the successful 1.58 bits model generate nonsensical output despite multiple attempts.
> 
> <img alt="Image" width="577" src="https://private-user-images.githubusercontent.com/3134202/407746029-d68bf7fa-22f5-4845-9ed9-1d39cf51b98a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDEwODQ2NjAsIm5iZiI6MTc0MTA4NDM2MCwicGF0aCI6Ii8zMTM0MjAyLzQwNzc0NjAyOS1kNjhiZjdmYS0yMmY1LTQ4NDUtOWVkOS0xZDM5Y2Y1MWI5OGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMDRUMTAzMjQwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDlkODViYTBjMTE1YWEwMDBlODMyMTFlMDY0MTgwMzAxZWQzOWFlMTBlZTQ3MWVjOGUzNDQ3M2IyY2I4YzI3YyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.uuXCaSJyOPCy7oKg7LDjMPBliG35hF_oIgGuKmAJrME">

I face the smiliar issue, the the model generate nonsensical output. Even that there is only one .gguf file donwloaded. May I know if your problem had been solved? Thanks

![Image](https://github.com/user-attachments/assets/31e94c61-96ee-4f5b-bc40-cb686c82c75b)

**Comment by PaulGilmartin** [comment]
I have an issue when trying to run `./llama-gguf-split`. I downloaded llama.cpp according to these [instructions](https://github.com/ggml-org/llama.cpp/blob/master/docs/build.md):

```
git clone https://github.com/ggml-org/llama.cpp.
cd llama.cpp/
cmake -B build
cmake --build build --config Release
```

But when I try to run./llama-gguf-split I get
```
./llama-gguf-split
-bash: ./llama-gguf-split: No such file or directory
```

Does anyone know where I'm going wrong?

**Comment by PaulGilmartin** [comment]
Hi all,

I am attempting to merge the gguf files from a hugging face DeepSeek V3 download. Using llama-gguf-split as follows:
```
llama.cpp/llama-gguf-split --merge DeepSeek-V3-GGUF/DeepSeek-V3-Q2_K_L/DeepSeek-V3-Q2_K_L-00001-of-00005.gguf DeepSeek-V3-GGUF/DeepSeek-V3-Q2_K_L/temp-merged.gguf
```
I encounter the following error:
```
gguf_merge: reading metadata DeepSeek-V3-GGUF/DeepSeek-V3-Q2_K_L/DeepSeek-V3-Q2_K_L-00001-of-00005.gguf ...gguf_init_from_file_impl: invalid magic characters: 'vers', expected 'GGUF'

gguf_merge:  failed to load input GGUF from DeepSeek-V3-GGUF/DeepSeek-V3-Q2_K_L/DeepSeek-V3-Q2_K_L-00001-of-00005.gguf
```

The files were downloaded by cloning https://huggingface.co/unsloth/DeepSeek-V3-GGUF. This is the content of the DeepSeek-V3-GGUF/DeepSeek-V3-Q2_K_L folder:

```
~/DeepSeek-V3-GGUF/DeepSeek-V3-Q2_K_L # ls
DeepSeek-V3-Q2_K_L-00001-of-00005.gguf
DeepSeek-V3-Q2_K_L-00002-of-00005.gguf
DeepSeek-V3-Q2_K_L-00003-of-00005.gguf
DeepSeek-V3-Q2_K_L-00004-of-00005.gguf
DeepSeek-V3-Q2_K_L-00005-of-00005.gguf
```

Does anyone know what's going wrong here/ how to fix this? Thanks in advance!

**Comment by MrEdigital** [comment]
More and more models are going the sharded route, exclusively.  This needs to be addressed soon.

**Comment by mcDandy** [comment]
It is highly needed. I want to use a custom gemma 3 but vision tower is separate. Llama.cpp does not help to merge vision into the model.

**Comment by 1472583610** [comment]
I plus-one this. Ollama needs to support this natively. Aside from it being an unreasonably large amount of work to constantly download and manually merge models, there seem to be issues when running the merged models on multiple GPUs.

We have a 2x A6000 AI server and it doesn't load merged models larger than what fits on a single card (48 GB).

Support for sharded models is becoming a must.

**Comment by mNandhu** [comment]
+1 - There's no mention about the incompatibility in ModelFile Docs either

**Comment by lknight** [comment]
+1 Its almost impossible to do something with DeepSeek (multiple files model) in Ollama and multiple A6000.

**Comment by LastMinuteStudio** [comment]
Relatively new to LLMs so pardon my ignorance. I've tried merging and running the split GGUFs especially on the latest R1 from unsloth https://huggingface.co/unsloth/DeepSeek-R1-0528-GGUF/tree/main/UD-IQ1_S. 

The way I run it is through their recommended command
`./llama-server --port 10000 --ctx-size 1024 --n-gpu-layers 40 --model a:/Deepseek-R1-0528-UD-IQ1_S-Merged.gguf`
`./llama-server --port 10000 --ctx-size 1024 --n-gpu-layers 40 --model a:/DeepSeek-R1-0528-UD-IQ1_S-00001-of-00004.gguf`

They work both merged and split in llama.cpp using llama-server.exe. 
I've then tried creating a Modelfile for the merged gguf in ollama
```
FROM a:\Deepseek-R1-0528-UD-IQ1_S-Merged.gguf
TEMPLATE """{{- if .System }}{{ .System }}{{ end }}
{{- range $i, $_ := .Messages }}
{{- $last := eq (len (slice $.Messages $i)) 1}}
{{- if eq .Role "user" }}<｜User｜>{{ .Content }}
{{- else if eq .Role "assistant" }}<｜Assistant｜>
  {{- if and $.IsThinkSet (and $last .Thinking) -}}
<think>
{{ .Thinking }}
</think>
{{- end }}{{ .Content }}{{- if not $last }}<｜end▁of▁sentence｜>{{- end }}
{{- end }}
{{- if and $last (ne .Role "assistant") }}<｜Assistant｜>
{{- if and $.IsThinkSet (not $.Think) -}}
<think>

</think>

{{ end }}
{{- end -}}
{{- end }}"""
PARAMETER min_p 0.01
PARAMETER repeat_penalty 1
PARAMETER top_p 0.95
PARAMETER num_predict 16384
PARAMETER num_ctx 1024
PARAMETER num_gpu 40
PARAMETER stop <｜begin▁of▁sentence｜>
PARAMETER stop <｜end▁of▁sentence｜>
PARAMETER stop <｜User｜>
PARAMETER stop <｜Assistant｜>
PARAMETER temperature 0.6
```

I create the model using  `ollama create DeepseekR1 -f Modelfile`
However when I run it through ollama, I keep getting an error telling me that the memory required is insufficient even though llama.cpp has no issues with loading it slowly. 
`Error: model requires more system memory (164.9 GiB) than is available (94.4 GiB)`
Is this due to the gguf being split or some other reason?  The merged gguf works fine in llama.cpp though

edit: I've increased my swap size just to get past the memory error but now I get the following
`llama runner process has terminated: error loading model: unable to allocate CUDA0 buffer`

**Comment by MrEdigital** [comment]
This doesn't appear to be given the weight it deserves.  It's now been more than a year since this was raised.

**Comment by giorgostheo** [comment]
This needs to be prioritized IMO. Shared GGUFs are the norm now. No support for them in probably the most used platform for local LLMs is bonkers...

**Comment by kappa8219** [comment]
Qwen Coder is coming :) Also sharded...

**Comment by giorgostheo** [comment]
I think its fair to say that this should become priority No.1 for the dev team at this point. Not having shared GGUF support will soon make ollama unusable...

**Comment by tolysz** [comment]
If `llama.cpp` has `llama_model_load_from_splits` ~we are almost half way... I wonder what is the desired solution... currently all the models are stored with hashed names, naive implementation could have some `virtual` to the top level folder and inside all files symlinked to the hashed files, filenames should follow the splits pattern... otherwise, the model could accept the list of splits and skip the symlinking~
edit:
We just need to provide all the hashed files as list... the filename itself is irrelevant...
```
    // Load the model from a file
    // If the file is split into multiple parts, the file name must follow this pattern: <name>-%05d-of-%05d.gguf
    // If the split file name does not follow this pattern, use llama_model_load_from_splits
    LLAMA_API struct llama_model * llama_model_load_from_file(
                             const char * path_model,
              struct llama_model_params   params);

    // Load the model from multiple splits (support custom naming scheme)
    // The paths must be in the correct order
    LLAMA_API struct llama_model * llama_model_load_from_splits(
                             const char ** paths,
                                 size_t    n_paths,
              struct llama_model_params    params);
```

**Comment by xNefas** [comment]
Just gonna add to the "noise" and say this should be a priority, it's really rough being unable to load sharded GGUFs with Ollama.

**Comment by Likkkez** [comment]
Is it really this hard to just add two files together automatically? Is this one of those delusional ideological stances or what?

**Comment by giorgostheo** [comment]
Honestly at this point Im not even sure this will ever be implemented. After all the fuss with GPT-OSS and the push for ollama turbo, it seems that this will be another one of those os projects that remains os mostly for show... I really hope Im wrong but tbh Im already transitioning to llama-cpp cause of this. I suggest that others do the same.

**Comment by OdinVex** [comment]
> Honestly at this point Im not even sure this will ever be implemented. After all the fuss with GPT-OSS and the push for ollama turbo, it seems that this will be another one of those os projects that remains os mostly for show... I really hope Im wrong but tbh Im already transitioning to llama-cpp cause of this. I suggest that others do the same.

? Turbo? What is that, some upsell or something? Is it time to fork Ollama?

**Comment by kappa8219** [comment]
"The Little Engine That Could" (c)

**Comment by bokkob556644-coder** [comment]
ollama run hf.co/asdgad/Qwen3-Coder-30B-A3B-Instruct-Q4_K_M-GGUF:Q4_K_M

https://huggingface.co/asdgad/Qwen3-Coder-30B-A3B-Instruct-Q4_K_M-GGUF

Error: pull model manifest: 400: {"error":"The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: https://github.com/ollama/ollama/issues/5245"}

**Comment by kappa8219** [comment]
> ollama run hf.co/asdgad/Qwen3-Coder-30B-A3B-Instruct-Q4_K_M-GGUF:Q4_K_M
> 
> https://huggingface.co/asdgad/Qwen3-Coder-30B-A3B-Instruct-Q4_K_M-GGUF
> 
> Error: pull model manifest: 400: {"error":"The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: [https://github.com/ollama/ollama/issues/5245"}](https://github.com/ollama/ollama/issues/5245%22%7D)

Welcome to the club

**Comment by SvenMeyer** [comment]
```bash
$ ollama pull hf.co/unsloth/MiniMax-M2-GGUF:Q3_K_XL
pulling manifest 
Error: pull model manifest: 400: {"error":"The specified tag is a sharded GGUF. Ollama does not support this yet. Please use another tag or \"latest\". Follow this issue for more info: https://github.com/ollama/ollama/issues/5245"}
```

nobody looking into this ?

**Comment by slenderq** [comment]
```
ollama run hf.co/unsloth/Kimi-K2-Thinking-GGUF:IQ1_S
pulling manifest
Error: pull model manifest: 400: {"error":"The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: https://github.com/ollama/ollama/issues/5245"}
```
It would be great to run this model in ollama!

**Comment by cvrunmin** [comment]
not a golang user tho, from my glance on the code, there is only one place which actually call model load function using llama.cpp here:
https://github.com/ollama/ollama/blob/8a75d8b0154511d2bafe16f230e9268ee7a511da/llama/llama.go#L259-L309
at L303, `C.llama_model_load_from_file` is called to load model, which the implementation is here:
https://github.com/ollama/ollama/blob/8a75d8b0154511d2bafe16f230e9268ee7a511da/llama/llama.cpp/src/llama.cpp#L304-L325
Yes, there is a function named `llama_model_load_from_splits` that should be able to load split files!
This function requires a list of split file paths as the parameter. Thus we have to know the correct order of the model split files (assume the function that actually load them don't guess how they are split). We might need to add fields in Modelfile to provide this information. This information should be provided by HF too when ollama is ready for multifile gguf (they only provide a list of hash-named blob with blob type afaik)

**Comment by tolysz** [comment]
The challenge is in the config files, as the files are renamed to some-hash... the config file needs to support storing a list of the filenames... like the type of file is not a string but a list of them.

**Comment by FearL0rd** [comment]
Looks like Ollama has more focus on their cloud instead of working on this

**Comment by OdinVex** [comment]
So it seems the only thing holding it back is a config-representation of split-files and a simple 'if split call load-splits' instead?

**Comment by rdeforest** [comment]
> Looks like Ollama has more focus on their cloud instead of working on this

One of the many great things about open-source projects is that everyone gets to work on whatever they want to work on. I bet if you put together a quality PR to address this issue, the team would consider merging it. Or if you don't want to wait you could just maintain your own fork.

If you don't want to help, that's fine too. Just don't complain about the priorities of volunteers please?

**Comment by OdinVex** [comment]
I think extension of Ollama manifests to describe split GGUFs (and their order) is necessary, first. Perhaps any Ollama developer could chime in for that?

**Comment by Mikec78660** [comment]
EDIT: Leaving this in case anyone else has this problem. Seems from the bash script that was posted you can then run:
`ollama create GLM-4.5-Air-Q4_0.gguf -f GLM-4.5-Air-Q4_0.gguf.model`
And voilà, the model is showing up in ollama, no need to use the import in openweui which doesn't seem to work.
*******

I use the llama.cpp method to combine my model:

```
GLM-4.5-Air-Q4_0-00001-of-00002.gguf
GLM-4.5-Air-Q4_0-00002-of-00002.gguf
```
And it seemed to work:

```
./llama-gguf-split --merge /mnt/AI/GLM-4.5-Air-Q4_0-00001-of-00002.gguf /mnt/AI/GLM-4.5-Air-Q4_0.gguf
gguf_merge: /mnt/AI/GLM-4.5-Air-Q4_0-00001-of-00002.gguf -> /mnt/AI/GLM-4.5-Air-Q4_0.gguf
gguf_merge: reading metadata /mnt/AI/GLM-4.5-Air-Q4_0-00001-of-00002.gguf done
gguf_merge: reading metadata /mnt/AI/GLM-4.5-Air-Q4_0-00002-of-00002.gguf done
gguf_merge: writing tensors /mnt/AI/GLM-4.5-Air-Q4_0-00001-of-00002.gguf done
gguf_merge: writing tensors /mnt/AI/GLM-4.5-Air-Q4_0-00002-of-00002.gguf done
gguf_merge: /mnt/AI/GLM-4.5-Air-Q4_0.gguf merged from 2 split with 803 tensors.
```

But trying to import "GLM-4.5-Air-Q4_0.gguf" into ollama after a minute or so I get an error saying error parsing the body. Any idea what I am doing wrong?

**Comment by OdinVex** [comment]
> EDIT: Leaving this in case anyone else has this problem. Seems from the bash script that was posted you can then run: `ollama create GLM-4.5-Air-Q4_0.gguf -f GLM-4.5-Air-Q4_0.gguf.model` And voilà, the model is showing up in ollama, no need to use the import in openweui which doesn't seem to work.
> 
> I use the llama.cpp method to combine my model:
> 
> ```
> GLM-4.5-Air-Q4_0-00001-of-00002.gguf
> GLM-4.5-Air-Q4_0-00002-of-00002.gguf
> ```
> 
> And it seemed to work:
> 
> ```
> ./llama-gguf-split --merge /mnt/AI/GLM-4.5-Air-Q4_0-00001-of-00002.gguf /mnt/AI/GLM-4.5-Air-Q4_0.gguf
> gguf_merge: /mnt/AI/GLM-4.5-Air-Q4_0-00001-of-00002.gguf -> /mnt/AI/GLM-4.5-Air-Q4_0.gguf
> gguf_merge: reading metadata /mnt/AI/GLM-4.5-Air-Q4_0-00001-of-00002.gguf done
> gguf_merge: reading metadata /mnt/AI/GLM-4.5-Air-Q4_0-00002-of-00002.gguf done
> gguf_merge: writing tensors /mnt/AI/GLM-4.5-Air-Q4_0-00001-of-00002.gguf done
> gguf_merge: writing tensors /mnt/AI/GLM-4.5-Air-Q4_0-00002-of-00002.gguf done
> gguf_merge: /mnt/AI/GLM-4.5-Air-Q4_0.gguf merged from 2 split with 803 tensors.
> ```
> 
> But trying to import "GLM-4.5-Air-Q4_0.gguf" into ollama after a minute or so I get an error saying error parsing the body. Any idea what I am doing wrong?

This Issue is about importing shard GGUF files, not about llama.cpp. But on a side note, I've never had llama.cpp produce a file that didn't end up in gibberish or corrupt output.

**Comment by shimmyshimmer** [comment]
> ```
> ollama run hf.co/unsloth/Kimi-K2-Thinking-GGUF:IQ1_S
> pulling manifest
> Error: pull model manifest: 400: {"error":"The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: https://github.com/ollama/ollama/issues/5245"}
> ```
> 
> It would be great to run this model in ollama!

We make smaller versions specifically for Ollama, see: https://huggingface.co/unsloth/MiniMax-M2-GGUF/blob/main/MiniMax-M2-UD-TQ1_0.gguf

Usually we do these non-sharded files for any model under 300B parameters or so. But it is very small and 1.77-bit ish

**Comment by cvrunmin** [comment]
> not a golang user tho, from my glance on the code, there is only one place which actually call model load function using llama.cpp here:
> 
> [ollama/llama/llama.go](https://github.com/ollama/ollama/blob/8a75d8b0154511d2bafe16f230e9268ee7a511da/llama/llama.go#L259-L309)
> 
> Lines 259 to 309 in [8a75d8b](/ollama/ollama/commit/8a75d8b0154511d2bafe16f230e9268ee7a511da)
> 
>  func LoadModelFromFile(modelPath string, params ModelParams) (*Model, error) { 
>  	cparams := C.llama_model_default_params() 
>  	cparams.n_gpu_layers = C.int(params.NumGpuLayers) 
>  	cparams.main_gpu = C.int32_t(params.MainGpu) 
>  	cparams.use_mmap = C.bool(params.UseMmap) 
>  	cparams.vocab_only = C.bool(params.VocabOnly) 
>   
>  	var devices []C.ggml_backend_dev_t 
>  	for _, llamaID := range params.Devices { 
>  		devices = append(devices, C.ggml_backend_dev_get(C.size_t(llamaID))) 
>  	} 
>  	if len(devices) > 0 { 
>  		devices = append(devices, C.ggml_backend_dev_t(C.NULL)) 
>  		devicesData := &devices[0] 
>   
>  		var devicesPin runtime.Pinner 
>  		devicesPin.Pin(devicesData) 
>  		defer devicesPin.Unpin() 
>   
>  		cparams.devices = devicesData 
>  	} 
>   
>  	if len(params.TensorSplit) > 0 { 
>  		tensorSplitData := &params.TensorSplit[0] 
>   
>  		var tensorSplitPin runtime.Pinner 
>  		tensorSplitPin.Pin(tensorSplitData) 
>  		defer tensorSplitPin.Unpin() 
>   
>  		cparams.tensor_split = (*C.float)(unsafe.Pointer(tensorSplitData)) 
>  	} 
>   
>  	if params.Progress != nil { 
>  		handle := cgo.NewHandle(params.Progress) 
>  		defer handle.Delete() 
>   
>  		var handlePin runtime.Pinner 
>  		handlePin.Pin(&handle) 
>  		defer handlePin.Unpin() 
>   
>  		cparams.progress_callback = C.llama_progress_callback(C.llamaProgressCallback) 
>  		cparams.progress_callback_user_data = unsafe.Pointer(&handle) 
>  	} 
>   
>  	m := Model{c: C.llama_model_load_from_file(C.CString(modelPath), cparams)} 
>  	if m.c == nil { 
>  		return nil, fmt.Errorf("unable to load model: %s", modelPath) 
>  	} 
>   
>  	return &m, nil 
>  } 
> 
> at L303, `C.llama_model_load_from_file` is called to load model, which the implementation is here:
> [ollama/llama/llama.cpp/src/llama.cpp](https://github.com/ollama/ollama/blob/8a75d8b0154511d2bafe16f230e9268ee7a511da/llama/llama.cpp/src/llama.cpp#L304-L325)
> 
> Lines 304 to 325 in [8a75d8b](/ollama/ollama/commit/8a75d8b0154511d2bafe16f230e9268ee7a511da)
> 
>  struct llama_model * llama_model_load_from_file( 
>          const char * path_model, 
>          struct llama_model_params params) { 
>      std::vector<std::string> splits = {}; 
>      return llama_model_load_from_file_impl(path_model, splits, params); 
>  } 
>   
>  struct llama_model * llama_model_load_from_splits( 
>          const char ** paths, 
>          size_t n_paths, 
>          struct llama_model_params params) { 
>      std::vector<std::string> splits; 
>      if (n_paths == 0) { 
>          LLAMA_LOG_ERROR("%s: list of splits is empty\n", __func__); 
>          return nullptr; 
>      } 
>      splits.reserve(n_paths); 
>      for (size_t i = 0; i < n_paths; ++i) { 
>          splits.push_back(paths[i]); 
>      } 
>      return llama_model_load_from_file_impl(splits.front(), splits, params); 
>  } 
> 
> Yes, there is a function named `llama_model_load_from_splits` that should be able to load split files!
> This function requires a list of split file paths as the parameter. Thus we have to know the correct order of the model split files (assume the function that actually load them don't guess how they are split). We might need to add fields in Modelfile to provide this information. This information should be provided by HF too when ollama is ready for multifile gguf (they only provide a list of hash-named blob with blob type afaik)

I only focused on llamarunner and didn't recognize that ollama has its own runner (ollamarunner) which is a different story. This thing will load models at `ml/backend/ggml/ggml.go` which really not supported multi-file gguf.
Anyway the modification of config format to support file path of sharded model in correct order is still necessary.

**Comment by giorgostheo** [comment]
> > ```
> > ollama run hf.co/unsloth/Kimi-K2-Thinking-GGUF:IQ1_S
> > pulling manifest
> > Error: pull model manifest: 400: {"error":"The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: https://github.com/ollama/ollama/issues/5245"}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It would be great to run this model in ollama!
> 
> We make smaller versions specifically for Ollama, see: https://huggingface.co/unsloth/MiniMax-M2-GGUF/blob/main/MiniMax-M2-UD-TQ1_0.gguf
> 
> Usually we do these non-sharded files for any model under 300M parameters or so. But it is very small and 1.77-bit ish

Hi Michael. Since ollama teams seems to really not care about the shared GGUF thing, it would be great if we got more of those "merged" exports for larger quants. For GLM4.6 for example, something like Q3 would be great. I understand that it is stretching it size-wise, but for us ollama users its the only way to go for now.

Thanks for all your work.

**Comment by OdinVex** [comment]
> ... This information should be provided by HF too when ollama is ready for multifile gguf (they only provide a list of hash-named blob with blob type afaik)

To the best of my knowledge splits always have their filenames suffixed (before extension) with a format of `splitNumber-totalSplits`. That's probably the only assumption that could be made about order. Maybe the backend doesn't care about order and loads them fine.

**Comment by cvrunmin** [comment]
> > ... This information should be provided by HF too when ollama is ready for multifile gguf (they only provide a list of hash-named blob with blob type afaik)
> 
> To the best of my knowledge splits always have their filenames suffixed (before extension) with a format of `splitNumber-totalSplits`. That's probably the only assumption that could be made about order. Maybe the backend doesn't care about order and loads them fine.

If the multi-file GGUF model is created by user using `ollama create -f Modelfile` and the filenames of split ggufs are nicely named as `xxxxxx-00001-of-00003.gguf` then this is what your case. However, when the model is pulled from Internet, we only have the hash of the file.
For example, this is the manifest of gpt-oss hosted on ollama registry (`https://registry.ollama.ai/v2/library/gpt-oss/manifests/latest`):
```json
{
  "schemaVersion":2,
  "mediaType":"application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType":"application/vnd.docker.container.image.v1+json",
    "digest":"sha256:776beb3adb235076157cfea408b8ea2a2d25eae99d7f5da997f607f6b69fa0fa",
    "size":489
  },
  "layers":[
    {
      "mediaType":"application/vnd.ollama.image.model",
      "digest":"sha256:e7b273f9636059a689e3ddcab3716e4f65abe0143ac978e46673ad0e52d09efb",
      "size":13793422144
    },
    {
      "mediaType":"application/vnd.ollama.image.template",
      "digest":"sha256:fa6710a93d78da62641e192361344be7a8c0a1c3737f139cf89f20ce1626b99c",
      "size":7240
    },
    {
      "mediaType":"application/vnd.ollama.image.license",
      "digest":"sha256:f60356777647e927149cbd4c0ec1314a90caba9400ad205ddc4ce47ed001c2d6",
      "size":11353
    },
    {
      "mediaType":"application/vnd.ollama.image.params",
      "digest":"sha256:d8ba2f9a17b3bbdeb5690efaa409b3fcb0b56296a777c7a69c78aa33bbddf182",
      "size":18
    }
  ]
}
```
In GGUF metadata of split GGUF, we have the split file information `split.no`, `split.tensors.count` and `split.count`. This is where llama.cpp check if the split file is provided in order:
https://github.com/ollama/ollama/blob/584e2d646fb4d2f1643b4da81a096d01114f5b2b/llama/llama.cpp/src/llama-model-loader.cpp#L526-L573
In worst case scenario, we can cache the ordering from the metadata when the model is first created or pulled, then changes towards the config spec could be minimal.

**Comment by shimmyshimmer** [comment]
> > > ```
> > > ollama run hf.co/unsloth/Kimi-K2-Thinking-GGUF:IQ1_S
> > > pulling manifest
> > > Error: pull model manifest: 400: {"error":"The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: https://github.com/ollama/ollama/issues/5245"}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > It would be great to run this model in ollama!
> > 
> > 
> > We make smaller versions specifically for Ollama, see: https://huggingface.co/unsloth/MiniMax-M2-GGUF/blob/main/MiniMax-M2-UD-TQ1_0.gguf
> > Usually we do these non-sharded files for any model under 300M parameters or so. But it is very small and 1.77-bit ish
> 
> Hi Michael. Since ollama teams seems to really not care about the shared GGUF thing, it would be great if we got more of those "merged" exports for larger quants. For GLM4.6 for example, something like Q3 would be great. I understand that it is stretching it size-wise, but for us ollama users its the only way to go for now.
> 
> Thanks for all your work.

Even though this is possible it might not be the best idea because if one file breaks or the internet gets cut off, you'll need to redownload the hundreds of GB again. It might be fine if you have good internet but around 50% of people have very slow internet :( But we'll see what we can do - it will be confusing for users to navigate which is which

**Comment by eurekin** [comment]
This still a thing?

Happy birthday to the issue I guess

**Comment by FearL0rd** [comment]
> This still a thing?
> 
> Happy birthday to the issue I guess

Now all the focus is on cloud

**Comment by johnml1135** [comment]
I had a similar issue so I spun up my own tooling to adapt poor-fitting GGUF models into ollama by reworking the top layers - https://github.com/johnml1135/ollama-copilot-fixer.

**Comment by giorgostheo** [comment]
> > > > ```
> > > > ollama run hf.co/unsloth/Kimi-K2-Thinking-GGUF:IQ1_S
> > > > pulling manifest
> > > > Error: pull model manifest: 400: {"error":"The specified repository contains sharded GGUF. Ollama does not support this yet. Follow this issue for more info: https://github.com/ollama/ollama/issues/5245"}
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > It would be great to run this model in ollama!
> > > 
> > > 
> > > We make smaller versions specifically for Ollama, see: https://huggingface.co/unsloth/MiniMax-M2-GGUF/blob/main/MiniMax-M2-UD-TQ1_0.gguf
> > > Usually we do these non-sharded files for any model under 300M parameters or so. But it is very small and 1.77-bit ish
> > 
> > Hi Michael. Since ollama teams seems to really not care about the shared GGUF thing, it would be great if we got more of those "merged" exports for larger quants. For GLM4.6 for example, something like Q3 would be great. I understand that it is stretching it size-wise, but for us ollama users its the only way to go for now.
> > 
> > Thanks for all your work.
> 
> Even though this is possible it might not be the best idea because if one file breaks or the internet gets cut off, you'll need to redownload the hundreds of GB again. It might be fine if you have good internet but around 50% of people have very slow internet :( But we'll see what we can do - it will be confusing for users to navigate which is which

Hey,

With GLM4.7 out, would it be possible that you upload a single gguf for the main config (q4_K_M or whatever it is)? You can add some sort of id like "mono" or "single" to make sure that users are not confused. I know it's not pretty, but it's an easy way to solve the complete lack of compatibility with ollama and allow tons more to use the newest and best models! 

Keep up the awesome work.

**Comment by scorpion7slayer** [comment]
I have the problem with kimi k2.5 will this be added in the future?

**Comment by boomam** [comment]
Its amusing that they went to the effort of changing the output of the Ollama error to reference this issue. :-p

**Comment by cvrunmin** [comment]
> Its amusing that they went to the effort of changing the output of the Ollama error to reference this issue. :-p

While such error messages is more likely provided from Huggingface side and not from ollama itself (same error message can be triggered by trying to access the manifest file of the model in browsers), it is still very amusing that more issues have been marked duplicate of this issue, meaning that some contributors know that this issue exists, but an pull request that claims that can solve this issues is now about four months old with zero comments from any contributors. Neither "this solution looks good!", nor "this solution is not good".

**Comment by OdinVex** [comment]
Anyone know of an alternative to Ollama?

**Comment by SvenMeyer** [comment]
@OdinVex I switched to LMstudio which does not have this problem and has a nice GUI as well.
Actually, I would not be surprised if you would find it much better in every respect. Also LMstudio contines to provide a solid basis to run AI models locally which is the whole point of ollama/LMstusio, while ollama now diverted to become just a proxy to online AI models.

**Comment by OdinVex** [comment]
> [@OdinVex](https://github.com/OdinVex) I switched to LMstudio which does not have this problem and has a nice GUI as well. Actually, I would not be surprised if you would find it much better in every respect. Also LMstudio contines to provide a solid basis to run AI models locally which is the whole point of ollama/LMstusio, while ollama now diverted to become just a proxy to online AI models.

Doesn't appear at all to be an alternative, though. Ollama's use at the moment is container-supported for network-based interactions.

**Comment by elkay** [comment]
How is this an issue still 2 years later? Isn't it as simple as combining the split files and using the single file after download? I know you can use llama.cpp to combine the files if you manually download them, but it's really unclear how you would then manually add that file into ollama manually. Makes no sense why the Ollama team is dragging their feet on just supporting the combine in the internal download process itself.

**Comment by FearL0rd** [comment]
> How is this an issue still 2 years later? Isn't it as simple as combining the split files and using the single file after download? I know you can use llama.cpp to combine the files if you manually download them, but it's really unclear how you would then manually add that file into ollama manually. Makes no sense why the Ollama team is dragging their feet on just supporting the combine in the internal download process itself.

looks like the focus today is Ollama Cloud

**Comment by alexanderjacuna** [comment]
Running into this issue as well with: hf.co/unsloth/Qwen3-Coder-Next-GGUF:UD-Q8_K_XL

This bug has been open since June of 2024, but lets put in a error message that references this issue with no traction after 2 years.

**Comment by SvenMeyer** [comment]
I found a solution and it is pretty easy, also adds at lot of other features and usability at the same time : use LMstudio.

**Comment by OdinVex** [comment]
> I found a solution and it is pretty easy, also adds at lot of other features and usability at the same time : use LMstudio.

Not a viable solution to those needing a drop-in replacement for software that specifically depends upon Ollama.

**Comment by alexanderjacuna** [comment]
> I found a solution and it is pretty easy, also adds at lot of other features and usability at the same time : use LMstudio.

My setup doesn't allow for this unfortunately.

**Comment by SvenMeyer** [comment]
@OdinVex @alexanderjacuna what software is so tightly coupled to ollama that you can not replace it with another inference service? At the end it should be just an IP and port and even that you could set the same way.
Also, if you prefer CLI and do not need the GUI, just use llama.cpp

**Comment by OdinVex** [comment]
> [@OdinVex](https://github.com/OdinVex) [@alexanderjacuna](https://github.com/alexanderjacuna) what software is so tightly coupled to ollama that you can not replace it with another inference service? At the end it should be just an IP and port and even that you could set the same way. Also, if you prefer CLI and do not need the GUI, just use llama.cpp

Several, but most commonly Open-WebUI.

**Comment by FearL0rd** [comment]
> > I found a solution and it is pretty easy, also adds at lot of other features and usability at the same time : use LMstudio.
> 
> Not a viable solution to those needing a drop-in replacement for software that specifically depends upon Ollama.

I have a drop-in Solution.
I've built a solution called Ovllm. it's essentially an Ollama-style wrapper, but for vLLM instead of llama.cpp. It's still a work in progress, but the core downloading feature is live. Instead of pulling from a custom registry, it downloads models directly from Hugging Face. Just make sure to set your HF_TOKEN environment variable with your API key. Check it out: https://github.com/FearL0rd/Ovllm
Ovllm is an Ollama-inspired wrapper designed to simplify working with vLLM.

**Comment by OdinVex** [comment]
> > > I found a solution and it is pretty easy, also adds at lot of other features and usability at the same time : use LMstudio.
> > 
> > 
> > Not a viable solution to those needing a drop-in replacement for software that specifically depends upon Ollama.
> 
> I have a drop-in Solution. I've built a solution called Ovllm. it's essentially an Ollama-style wrapper, but for vLLM instead of llama.cpp. It's still a work in progress, but the core downloading feature is live. Instead of pulling from a custom registry, it downloads models directly from Hugging Face. Just make sure to set your HF_TOKEN environment variable with your API key. Check it out: https://github.com/FearL0rd/Ovllm Ovllm is an Ollama-inspired wrapper designed to simplify working with vLLM.

If it doesn't have complete feature-parity and speak Ollama so other software would integrate then it's not a drop-in solution. Considering the README has enough spelling/grammar issues I'm gravely concerned it's AI, or at the very least unpolished. Good luck with your project, but it's not a drop-in solution.

Edit: Considering Ollama uses llama and llama supports shards I'd wager it'd be better to probably PR it (at least for now, until Ollama is forked by someone that cares about shard-support and more).

**Comment by FearL0rd** [comment]
> > > > I found a solution and it is pretty easy, also adds at lot of other features and usability at the same time : use LMstudio.
> > > 
> > > 
> > > Not a viable solution to those needing a drop-in replacement for software that specifically depends upon Ollama.
> > 
> > 
> > I have a drop-in Solution. I've built a solution called Ovllm. it's essentially an Ollama-style wrapper, but for vLLM instead of llama.cpp. It's still a work in progress, but the core downloading feature is live. Instead of pulling from a custom registry, it downloads models directly from Hugging Face. Just make sure to set your HF_TOKEN environment variable with your API key. Check it out: https://github.com/FearL0rd/Ovllm Ovllm is an Ollama-inspired wrapper designed to simplify working with vLLM.
> 
> If it doesn't have complete feature-parity and speak Ollama so other software would integrate then it's not a drop-in solution. Considering the README has enough spelling/grammar issues I'm gravely concerned it's AI, or at the very least unpolished. Good luck with your project, but it's not a drop-in solution.
> 
> Edit: Considering Ollama uses llama and llama supports shards I'd wager it'd be better to probably PR it (at least for now, until Ollama is forked by someone that cares about shard-support and more).

Thx. This is the first release, and it will become more mature over time. It works for my needs with OpenWebUI and custom apps. It also merges the .gguf (it works with safetensors also, only pass the hf location ie. google/gemma-7b-it)

**Comment by Mikec78660** [comment]
llama.cpp in router mode should works exactly like ollama now. and can use multi part gguf files.

**Comment by OdinVex** [comment]
> llama.cpp in router mode should works exactly like ollama now. and can use multi part gguf files.

So software like Open-WebUI can (without any changes except IP address and port) speak to it as if it were Ollama, even with the Ollama-specific code? And there's an official container for it as well? Not seeing it at all, so... Edit: See my lower post about how this went (failure, does not at all work like Ollama).

**Comment by Mikec78660** [comment]
@OdinVex  yes.
A very minimal implementation is:
`llama-server --host [0.0.0.0, or hostname] --port 8080 --models-dir /mnt/AI`
if you do this and create a connection in openwebui to [ip or dns name]:8080/v1, it will give you any model in the /mnt/AI directory as an option in openwebui.

Even better is using a config.ini file where you can set the setting for each model
`llama-server --host 0.0.0.0 --port 8080 --models-dir /mnt/AI --models-preset config.ini`
This will allow you to set a custom ctx size, kv cache setting, etc.

**Comment by OdinVex** [comment]
> [@OdinVex](https://github.com/OdinVex) yes. A very minimal implementation is: `llama-server --host [0.0.0.0, or hostname] --port 8080 --models-dir /mnt/AI` if you do this and create a connection in openwebui to [ip or dns name]:8080/v1, it will give you any model in the /mnt/AI directory as an option in openwebui.
> 
> Even better is using a config.ini file where you can set the setting for each model `llama-server --host 0.0.0.0 --port 8080 --models-dir /mnt/AI --models-preset config.ini` This will allow you to set a custom ctx size, kv cache setting, etc.

Edit: I see, you meant an OpenAI endpoint, not an *Ollama* endpoint. Will try and report back if it works.
Edit: It does *not* work, unfortunately. Trying to download just results in the API reporting 404.

**Comment by raro42** [comment]
.

**Comment by CleyFaye** [comment]
We already know about the issue, and the general outline of what should be done.

I don't see the value of regurgitating the existing discussion, especially to end on the suggestion to "do what was proposed, then tests and docs".

**Comment by raro42** [comment]
@CleyFaye sorry for disturbing. Will edit the comment. Thanks for commenting.

**Comment by lawcontinue** [comment]
We ran into this exact limitation while building [Hippo](https://github.com/lawcontinue/hippo), a lightweight local LLM manager that uses llama-cpp-python to load GGUF models.

**What we found:**

llama-cpp-python handles multi-file GGUF transparently — when you pass a path like `model-00001-of-00005.gguf`, it detects and loads all shards automatically. So the technical blocker isn't on the inference side; it's on Ollama's import/registry layer that expects a single file.

**Workaround we use in Hippo:**

For our users, we document a pre-import step: concatenate sharded GGUF files using `gguf-split --merge`, then point Hippo at the merged file. It works but adds friction — a 70B model merge takes 2-3 minutes and temporarily doubles disk usage.

**Implementation thoughts for Ollama:**

The cleanest path would be to extend `ollama create` to accept glob patterns or a directory:

```bash
# Current (single file)
ollama create mymodel -f Modelfile  # FROM ./model.gguf

# Proposed (multi-file)
ollama create mymodel -f Modelfile  # FROM ./model-*.gguf
```

Under the hood, this would pass the first shard path to the GGUF loader (which already handles multi-file), or merge them during the blob creation step.

**One consideration:** sharded files from HuggingFace often have naming conventions like `model-00001-of-00005.gguf`. Detecting and sorting these correctly matters — lexicographic sort on the counter portion, not the full filename.

Happy to share more details from our Hippo implementation if helpful.

**Comment by Phixsura** [comment]
Implemented and tested. PR: #16223

**What's covered:**

Multiple `FROM` lines in a Modelfile now work for split GGUF models:

```dockerfile
FROM /path/to/model-00001-of-00003.gguf
FROM /path/to/model-00002-of-00003.gguf
FROM /path/to/model-00003-of-00003.gguf
```

The order of `FROM` lines does not matter — shards are sorted by `split.no` metadata before being written to the manifest.

**Under the hood:** `llama_model_load_from_splits` is used instead of `llama_model_load_from_file` when multiple paths are present. `LoadModelFromFile` on a split shard triggers llama.cpp's filename-pattern auto-discovery, which fails on Ollama's content-addressed blob names.

**Tested with:**
- 2-shard model from `llama-gguf-split --split-max-size 50M` (SmolLM 135M, Metal — two separate buffer allocations confirmed in logs)
- 3-shard model with reversed `FROM` order — manifest verified to be in correct `split.no` order; inference succeeds
- Incomplete shard set (2 of 3) produces `split GGUF declares 3 shards but 2 were provided`

**Comment by bg4ezt** [comment]
@homeserver:~$ docker exec -it ollama_engine ollama run hf.co/unsloth/Qwen3.5-122B-A10B-GGUF:UD-Q8_K_XL
pulling manifest 
Error: pull model manifest: 400: {"error":"The specified tag is a sharded GGUF. Ollama does not support this yet. Please use another tag or \"latest\". Follow this issue for more info: https://github.com/ollama/ollama/issues/5245"}

**Comment by bg4ezt** [comment]
> > Just in case someone would find this issue, like I did a few weeks ago, without knowing any workaround. Currently probably one of the easiest ways import multifile gguf into Ollama would be to:
> > 
> > 1. Download pre-compiled binaries of llama.cpp: https://github.com/ggerganov/llama.cpp/releases (or install according to their manual)
> > 2. run command:
> >    `./llama-gguf-split --merge mymodel-00001-of-00002.gguf out_file_name.gguf`
> >    for example
> >    `./llama-gguf-split --merge Mistral-Large-Instruct-2407-IQ4_XS-00001-of-00002.gguf outfile.gguf`
> > 
> > Hope this will help somebody.
> 
> For anyone else wondering how this merges all the other parts of the model or if you have to run this command for each split of the file to merge it into the new one or something, this [reddit post](https://www.reddit.com/r/LocalLLaMA/comments/1cf6n18/comment/l1o0opp/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) laid it out. The gguf-split utility simply figures out the rest of the shards to merge together given the name of the first. I assume it requires they be names basically the same with only numbers deviating.

I have merged 6 files into 1GGUF with right model file but failed to use native function calling and vision. only text message works

**Comment by ErenAta16** [comment]
Picking up from @dhiltgen's clarification on the linked #14575 thread (2026-06-30): split GGUF is supported now, but only through `ollama create` - the `hf.co/...` registry path synthesizes a model on the fly and doesn't glue the shards together, which is why people keep hitting this from the `ollama run hf.co/...` direction even though the underlying capability landed.

Traced the working half to see what the gap actually needs. `server/create.go` already has everything required to merge shards: `splitGGUFName` parses the standard llama.cpp `name-NNNNN-of-NNNNN.gguf` pattern, `splitGGUFGroupKey` groups layers by that pattern cross-checked against the GGUF's own `split.count`/`split.no` metadata (so a mis-numbered or incomplete set fails loudly instead of silently merging wrong), and `mergeSplitGGUFLayers` does the actual join. That's exactly the logic the manual workaround in this thread's description recreates by hand: download the repo, `cd` into it, run `ollama create`.

The reason the `hf.co/` pull path can't just call that logic today is that it never gets a directory listing to detect the pattern from - it resolves one blob at a time against a specific tag/filename, not a whole repo. Hugging Face's own model API does expose the file listing for a repo (the `siblings` field on the model info endpoint), which is enough to detect a `-NNNNN-of-NNNNN.gguf` set before committing to a download. So the shape of a fix, without inventing new merge logic, would be: when resolving an `hf.co/org/repo:tag` reference, list the repo's files first, check whether the selected quant's GGUF is actually a shard set rather than a single file, and if so fetch all matching shards and hand them to the same `splitGGUFGroupKey`/`mergeSplitGGUFLayers` path `ollama create` already uses, instead of only ever expecting one file per tag.

That would remove the manual download-then-create workaround for the common case (someone just wants `ollama run hf.co/...` to work for a model that happens to be sharded on HF), while keeping the shard-count/metadata validation that already protects `ollama create` from being pointed at a mismatched or partial set.

**Comment by bokkob556644-coder** [comment]
thank 🧒

في السبت، ١١ يوليو ٢٠٢٦, ١٠:٥٣ م ErenAta16 ***@***.***> كتب:

**Comment by sbe-arg** [comment]
Not reviving but asking.

Can we "patch it" for convenience sake?

When the return is sharded then run something like the scripts already proposed a year ago. https://github.com/ollama/ollama/issues/5245#issuecomment-2610450171
With a user confirmation you proceed to auto download > create from source file? Y/N.
Then download the guufs and run ollama create.

It can even be conditional to having the hf cli installed or similar. 

🤷‍♂️

**Comment by ErenAta16** [comment]
@sbe-arg the shape of the patch matters here, and I think it rules out the version you are describing.

Re-checked the code since my July comment; nothing relevant moved. The merge helpers are still there and still wired up:

```
server/create.go:1111  splitGGUFName
server/create.go:1128  splitGGUFGroupKey
server/create.go:1155  mergeSplitGGUFLayers
server/create.go:456   mergeSplitGGUFLayers(splitGroups[key])   <- the only call site
```

That call site is on the `ollama create` path. So the capability exists locally and works on files you already have.

**The `hf.co/...` failure never reaches it, because it is not a local failure.** The error is

```
Error: pull model manifest: 400: {"error":"The specified tag is a sharded GGUF ...
```

and that string does not appear anywhere in this repository. It comes back from the registry when it resolves the tag, which means the refusal happens before a single byte is downloaded. There is no partially-fetched set of shards sitting on disk for a post-pull script to glue together, so "detect sharded, then run the merge" has nothing to run on.

That is why this keeps getting reported as unfixed even though the underlying capability landed: the two halves are on opposite sides of the network boundary.

A client-side patch is still possible, but it is a different thing from what was proposed a year ago. It would have to be a fallback: catch the 400, fetch the shard files from HuggingFace directly rather than through the registry proxy, and hand them to the existing `create` path. That is a new download route with its own auth, resume and cache-layout questions, not a wrapper around the merge that already exists.

The smaller fix is on the registry side: have it emit a manifest with one layer per shard instead of refusing the tag. Then nothing in the client changes at all, because `create.go:456` already handles a multi-layer split group, and `splitGGUFGroupKey` already cross-checks the GGUF's own `split.count` / `split.no` metadata so an incomplete set fails loudly rather than merging wrong.

Neither of those is something I can land from outside, but it is worth the thread being precise about which of the two is being asked for, since they have different owners.

**Comment by sbe-arg** [comment]
I thought we could pull using filter for files and not tags thats the main change. as a fallback.

**Comment by kappa8219** [comment]
Let's see what hf will say

**Comment by kappa8219** [comment]
Following up on [the analysis above](https://github.com/ollama/ollama/issues/5245#issuecomment-5241307168) — the call-graph half is right, but the conclusion that a registry-side manifest change needs no client change does not hold. I checked, pinned to `39df91c9826b3c0c83677f75cd230d8848d287c3`.

**The merge is reachable only from `create`, via explicitly supplied files:**

```
mergeSplitGGUFLayers            server/create.go:1155
  └─ called only at            server/create.go:456
       inside convertModelFromFilesWithMediaType   (declared :411)
         ├─ convertModelFromFiles                  (:403)
         │    ├─ create.go:192   r.Files
         │    └─ create.go:230   r.Adapters
         └─ convertDraftModelFromFiles             (:407)
              └─ create.go:215   r.DraftFiles
```

`server/images.go`, `server/download.go` and `server/routes.go` contain zero references to `mergeSplitGGUFLayers` or `splitGGUFGroupKey`, so `PullModel` cannot reach it by any path.

**And a per-shard manifest would not be silently handled — it would be silently wrong.** In `GetModel` (`server/images.go:670`):

```go
for _, layer := range mf.Layers {
    switch layer.MediaType {
    case "application/vnd.ollama.image.model":
        m.ModelPath = filename   // images.go:711
```

`ModelPath` is a single field assigned per matching layer, with no guard or dedupe — line 711 is its only assignment in `server/`, and nothing validates the number of `image.model` layers. A manifest with N shard layers therefore leaves `ModelPath` pointing at whichever shard was last, producing a model backed by a fraction of its tensors rather than a clean error.

So "emit one layer per shard and nothing in the client changes" would need, at minimum, merge-after-download on the pull path plus a guard rejecting multi-model-layer manifests on older clients. That is new logic, not a wrapper around the existing merge.

**The `create` path itself does work, end to end.** Verified on 0.32.9 with a deliberately split small model (415 MB q2_k, `llama-gguf-split --split-max-size 150M` into 3 shards):

```console
$ ollama create qwen-e2e-test
copying file sha256:0ff673cc…      # shard 1
copying file sha256:f10b12bb…      # shard 2
copying file sha256:0222b690…      # shard 3
parsing GGUF
verifying conversion
creating new layer sha256:7e851307…
writing manifest
success
```

Three shards in, one `image.model` layer out; the model loads and generates. The merged blob is exactly the same length as the original unsplit file (415182688 bytes) with a different hash, which is `removeSplitMetadata` (`create.go:996`) stripping `split.no` / `split.count` / `split.tensors.count`. An incomplete set fails loudly at `create.go:1165` (`has 1 shards, expected 3`) rather than merging wrong.

For reference, HF has said they will reword the registry 400 to point at this `create` workaround rather than at this issue: https://github.com/huggingface/hub-docs/issues/2702

**Comment by ErenAta16** [comment]
@kappa8219 you are right and I was wrong. I traced it independently at the same pin (`39df91c9`) rather than take the correction on trust, and I get the same answer. The "registry emits one layer per shard and nothing in the client changes" half of my earlier comment should be struck.

Both halves hold up:

```
convertModelFromFiles          create.go:403  <- r.Files      (:192)
                                              <- r.Adapters   (:230)
convertDraftModelFromFiles     create.go:407  <- r.DraftFiles (:215)
   both -> convertModelFromFilesWithMediaType (:411) -> mergeSplitGGUFLayers (:456)
```

`PullModel` is `images.go:991` and reaches none of it. The merge only ever sees files a create request handed it by name, which is exactly the thing a pull does not have.

**One detail worth adding, because it changes what the fix is.** I first read your `ModelPath` point as "the pull path needs a guard against multi-model-layer manifests." It is more than that. Look at the four path cases in the same switch:

```go
case "application/vnd.ollama.image.model":
    m.ModelPath = filename                                  // images.go:711
case manifest.MediaTypeImageDraft:
    m.DraftPath = filename                                  // :725
case "application/vnd.ollama.image.adapter":
    m.AdapterPaths = append(m.AdapterPaths, filename)       // :731
case "application/vnd.ollama.image.projector":
    m.ProjectorPaths = append(m.ProjectorPaths, filename)   // :733
```

Adapters and projectors accumulate because they are expected to be plural. Model and draft assign because the struct says there is exactly one, and `ModelPath` is a `string` at `images.go:68`, read at eight sites in `server/` alone and handed to the runner as `Args` at `images.go:584`.

So "one layer per shard" is not a manifest format this client under-validates, it is one it has no representation for. Supporting it means either merging on the download path so a manifest still yields a single blob, or making `ModelPath` plural through every consumer including the runner argument. The first is contained; the second is not, and neither is a guard.

Which I think makes the registry-side option strictly worse than I pitched it, not just unfinished. A per-shard manifest published today would be mishandled by every already-installed client, and no server-side change can reach those.

Good catch on the HF thread too. With Pierrci rewording the 400 to point at `ollama create`, the documented answer for users is settled and what is left in this issue is narrower than it has been: not "sharded GGUFs are unusable" but "pulling them by tag needs merge-on-download." Worth retitling around that, since the create workaround now has an official pointer at it and the thread has spent a year conflating the two.

**Comment by kappa8219** [comment]
For me ollama build from #17743 works fine with `pull` not create.

<img width="1658" height="288" alt="Image" src="https://github.com/user-attachments/assets/52d6c305-c215-4cbf-a644-3b6e3ff15edb" />

both for x86 and arm. I can publish dockerfiles if needed.
